lampe/inference.py

```diff
@@ -532,20 +546,198 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
+            the moments are used to standardize the parameters.
+        build: The network constructor (e.g. :class:`lampe.nn.ResMLP`).
+        kwargs: Keyword arguments passed to the constructor.
+    """
+
+    def __init__(
+        self,
+        theta_dim: int,
+        x_dim: int,
+        t_dim: int = 1,
+        moments: Tuple[Tensor, Tensor] = None,
+        build: Callable[[int, int], nn.Module] = MLP,
+        **kwargs,
+    ):
+        super().__init__()
+
+        if moments is None:
+            self.standardize = nn.Identity()
+        else:
+            mu, sigma = moments
+            self.standardize = Affine(-mu / sigma, 1 / sigma, trainable=False)
+
+        self.net = build(theta_dim + x_dim + t_dim * 2, theta_dim, **kwargs)
+
+        self.register_buffer('periods', torch.arange(t_dim) + 1)
+        self.register_buffer('zeros', torch.zeros(theta_dim))
+        self.register_buffer('ones', torch.ones(theta_dim))
+
+    def forward(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        r"""
+        Arguments:
+            theta: The parameters :math:`\theta`, with shape :math:`(*, D)`.
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+            t: The time :math:`t`, with shape :math:`(*,).`
+
+        Returns:
+            The rescaled score :math:`s_\phi(\theta, x, t)`, with shape :math:`(*, D)`.
+        """
+
+        t = self.periods * math.pi * t[..., None]
+        t = torch.cat((t.cos(), t.sin()), dim=-1)
+
+        theta = self.standardize(theta)
+        theta, x, t = broadcast(theta, x, t, ignore=1)
+
+        return self.net(torch.cat((theta, x, t), dim=-1))
+
+    def alpha(self, t: Tensor) -> Tensor:
+        return torch.exp(-8.0 * t**2)
+
+    def beta(self, t: Tensor) -> Tensor:
+        return 16.0 * t
+
+    def ode(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        drift = -self.beta(t) / 2 * theta
+        diffusion = -self.beta(t) / 2 * (1 + self.alpha(t)) * self.forward(theta, x, t)
+
+        return drift + diffusion
+
+    def flow(self, x: Tensor) -> Distribution:
+        r"""
+        Arguments:
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+
+        Returns:
+            The posterior distribution :math:`p_\phi(\theta | x)` induced by the
+            probability flow ODE.
+
+        Note:
+            The :func:`log_prob` method of the returned distribution is an unbiased
+            stochastic estimator of the true log-density. One should average over
+            a sufficient number of evaluations to attain small errors.
+        """
+
+        batch_shape = x.shape[:-1]
+
+        return NormalizingFlow(
+            transform=FFJTransform(
+                f=lambda theta, t: self.ode(theta, x, t),
+                time=x.new_tensor(1.0),
+                phi=(x, *self.parameters()),
+            ),
+            base=DiagNormal(self.zeros, self.ones).expand(batch_shape),
+        )
```

Comment:
After experimenting with this part of the code, it would be nice to have a unified API such that `nse.flow` can be passed directly to the coverage utils. This is currently not convenient because `nse.flow.log_prob` is a noisy estimate of the log-density.